clang/lib/CodeGen/Targets/RISCV.cpp

```diff
@@ -428,13 +444,11 @@ ABIArgInfo RISCVABIInfo::classifyArgumentType(QualType Ty, bool IsFixed,
   if (isEmptyRecord(getContext(), Ty, true) && Size == 0)
     return ABIArgInfo::getIgnore();
 
-  bool IsSingleCapRecord = false;
-  if (auto *RT = Ty->getAs<RecordType>())
-    IsSingleCapRecord = Size == getTarget().getCHERICapabilityWidth() &&
-                        getContext().containsCapabilities(RT->getDecl());
+  auto CapabilityWidth = getTarget().getCHERICapabilityWidth();
+  bool ForcePassInCapRegs = shouldPassStructDirectInCapRegisters(Size, Ty);
 
   bool IsCapability = Ty->isCHERICapabilityType(getContext()) ||
-                      IsSingleCapRecord;
+                      (ForcePassInCapRegs && Size == CapabilityWidth);
```

Comment:
This could be just `ForcePassInCapRegs` if you change `==` to `<=` in `shouldPassStructDirectInCapRegisters`

Reply:
Here's the relevant bit of the discussion: the final goal is that of having `IsCapability` true iff. the type is itself a capability or a struct made of a single capability, the case where the struct is composed of >= 2 capabilities should not entail `IsCapability == true`. 

Latest commit renames `IsCapability` to `IsSingleCapability`!